src/System.Net.Primitives/src/System/Net/IPEndPoint.cs

```diff
@@ -119,6 +119,157 @@ public int Port
+                            *      
+                            *      While we could look for a leading colon in that scenario, not all interop addresses begin with that:
+                            *      
+                            *          1::1.0.0.0  (valid)
+                            *      
+                            *      Other complications:
+                            *           Short IPv4 addresses:   0:1        (valid IPv4 with port)
+                            *           IPv4 expressed in hex:  0xFFFFFFFF (valid: 255.255.255.255)
+                            *           
+                            *      Instead of looking for IPv4, we're going to look for indications that this is IPv6. We'll start at the beginning 
+                            *      of the string and loop up to the spot where we saw the last colon. If we see another colon along the way, it's IPv6.
+                            */
+
+                            sliceLength = i;    // Assume it's IPv4 with a port
+                            for (int j = 0; j < i; j++)
+                            {
+                                if (endPointSpan[j] == ':') // No, it's IPv6 without a port
+                                {   
+                                    sliceLength = endPointSpan.Length;
+                                    break;
+                                }
+                            }
+                            
+                            // IPv4 without a port will run through the entire loop and exit without hiting a colon, so we're good in that case.
+                        }
+
+                        break;  // We can break out of the main loop now. We know everything we need to know.
+                    }
+                    else if ('0' <= digit && digit <= '9')
+                    {
+                        // -1 is used to indicate that we have not seen any numeric digits while searching for a port (e.g. 192.168.1.2:)
+                        // We do not want that to be a valid EndPoint.
+                        if (port == -1)
+                        {
+                            port = 0;
+                        }
+
+                        // We'll avoid overflow in cases where someone passes in garbage
+                        if (port < MaxPort)
+                        {
+                            port += multiplier * (digit - '0');
+                            multiplier *= 10;
+                        }
+                    }
+                    else
+                    {
+                        // If we see anything other than a numeric digit while processing the port number then we'll note that down for later.
+                        // We cannot determine the validity of that until we know for sure whether a port is actually present.
+                        encounteredInvalidPortCharacter = true;
+                    }
+                }
```

Comment:
I'm not understanding why all of this manual parsing/iterating is necessary.  Why can't it be something along the lines of (pseudo-code):
```C#
// Look for the last colon.  If we don't find one, it's at best an IPv4 address without a port.
// We can also simplify later cases by handling the case where we find a colon at the 0th
// place, in which case there's either an invalid address or this is an IPv6 address without
// a port, and we can just parse this as an IPAddress as well.
int lastColonPos = endPointSpan.LastIndexOf(':');
if (lastColonPos <= 1)
{
    ... // IPAddress.TryParse(endPointSpan, out address);
    return;
}

// Look to see if this is an IPv6 address with a port.
if (endPointSpan[lastColonPos-1] == ']')
{
    ... // IPAddress.TryParse(endPointSpan.Slice(0, lastColonPos - 2), out address);
    ... // int.TryParse(endPoint.Slice(lastColonPos+1), out port);
    return;
}

// This is either an IPv6 address without a port, or it's an IPv4 address with a port.
// IFF it's the former, there will be at least one more semicolon.
int secondToLastColonPos = endPointSpan.LastIndexOf(':', lastColonPos - 1);
if (secondToLastColonPos == -1)
{
   ... // IPAddress.TryParse(endPointSpan, out address);
   return;
}

... // IPAddress.TryParse(endPointSpan.Slice(0, lastColonPos-1), out address);
... // int.TryParse(endPointSpan.Slice(lastColonPos+1), out port);
return;
```
?

Reply:
Possibly. There's really not a lot to the actual parsing routine. My attempt was to handle it all as we go through (i.e. I'm looping through looking for a certain character anyway, might as well convert the port to a number while I'm at it), but it does become a bit convoluted that way. Would be easy enough to change

Reply:
Thanks. Unless there's a substantial performance difference, I'd rather keep the code as simple as possible and reuse existing parsers where we have them (e.g. uint.TryParse rather than parsing the digits manually).

Reply:
Simplified version checked in. I'm still trying to get a feel for the coding style of corefx. The original attempt kind of took `IPv4AddressHelper.ParseNonCanonical` as a source for what things should look like, where the method processed sections with loops while looking for specific character values. Not necessarily a defense of the first version, just an idea of where that came from.

https://github.com/dotnet/corefx/blob/09199cfcb7b55d71a040fad1fa3538d13e5f88f5/src/System.Net.Primitives/src/System/Net/IPv4AddressHelper.cs#L175